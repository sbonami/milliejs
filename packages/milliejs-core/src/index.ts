export type { Entity } from "./entity"
export { isEntity } from "./entity"
export type { Resource } from "./resource"

export type { LifecycleEventsType } from "./lifecycle"
export { LifecycleEvents } from "./lifecycle"
