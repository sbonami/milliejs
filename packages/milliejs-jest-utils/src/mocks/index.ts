export { makeMockEntity } from "./entity"
export { makeMockQuery } from "./query"
export { makeMockResource } from "./resource"
