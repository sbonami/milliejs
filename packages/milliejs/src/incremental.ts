import invariant from "tiny-invariant"
import type { Resource } from "@milliejs/core"
import type {
  PublisherActionInterface,
  SubscriberActionInterface,
} from "@milliejs/store-base"
import {
  isPublisherActionEventInterface,
  PublisherActionEventInterface,
  PublisherActionEventWrapper,
} from "./store/events"

export type StoreConstructorSourceOptions<R extends Resource> = {
  sourcePublisher?:
    | PublisherActionEventInterface<R>
    | PublisherActionInterface<R>
  sourceSubscriber?: SubscriberActionInterface
}

export class IncrementalStore<R extends Resource> {
  private _replicaStore?: PublisherActionEventInterface<R>
  private _sourcePublisher?: PublisherActionEventInterface<R>
  private _sourceSubscriber?: SubscriberActionInterface

  constructor(
    readonly resource: R,
    replicaStore:
      | PublisherActionEventInterface<R>
      | PublisherActionInterface<R>,
    sourceInterfaces?: StoreConstructorSourceOptions<R>,
  ) {
    this.replicaStore = replicaStore

    if (sourceInterfaces?.sourcePublisher)
      this.sourcePublisher = sourceInterfaces.sourcePublisher

    if (sourceInterfaces?.sourceSubscriber)
      this.sourceSubscriber = sourceInterfaces.sourceSubscriber
  }

  private set replicaStore(
    newReplicaStore:
      | PublisherActionEventInterface<R>
      | PublisherActionInterface<R>
      | undefined,
  ) {
    invariant(
      typeof newReplicaStore !== "undefined",
      "Replica store cannot be unset",
    )

    this._replicaStore = isPublisherActionEventInterface(newReplicaStore)
      ? newReplicaStore
      : new PublisherActionEventWrapper(newReplicaStore)
  }

  get replicaStore() {
    return this._replicaStore
  }

  set sourcePublisher(
    newSourcePublisher:
      | PublisherActionEventInterface<R>
      | PublisherActionInterface<R>
      | undefined,
  ) {
    if (typeof newSourcePublisher === "undefined") {
      this._sourcePublisher = undefined
    } else {
      this._sourcePublisher = isPublisherActionEventInterface(
        newSourcePublisher,
      )
        ? newSourcePublisher
        : new PublisherActionEventWrapper(newSourcePublisher)
    }
  }

  get sourcePublisher() {
    return this._sourcePublisher
  }

  set sourceSubscriber(
    newSourceSubscriber: SubscriberActionInterface | undefined,
  ) {
    if (typeof newSourceSubscriber === "undefined") {
      this._sourceSubscriber = undefined
    } else {
      this._sourceSubscriber = newSourceSubscriber
    }
  }

  get sourceSubscriber() {
    return this._sourceSubscriber
  }
}

export default IncrementalStore
