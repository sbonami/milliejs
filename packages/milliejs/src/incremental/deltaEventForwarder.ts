import invariant from "tiny-invariant"
import type { Entity, Resource } from "@milliejs/core"
import {
  LifecycleEvents,
  SubscriberActionInterface,
} from "@milliejs/store-base"
import type { IncrementalStore } from "../incremental"
import type { PublisherActionEventInterface } from "../store/events"

export const deltaEventForwarder = <R extends Resource = Resource>(
  incrementalStore: IncrementalStore<R>,
  store: PublisherActionEventInterface<R> | SubscriberActionInterface,
) => {
  store.on(LifecycleEvents.Save, (entity: Entity<R>) => {
    try {
      invariant(
        incrementalStore.resource.id === entity.resource.id,
        "Subscriber Resource is for another Incremental Store",
      )
      incrementalStore.emit(
        LifecycleEvents.Save,
        incrementalStore.resource,
        entity,
      )
      incrementalStore.emit(
        LifecycleEvents.Delta,
        incrementalStore.resource,
        entity,
      )
    } catch (error) {
      store.emit("error", error)
    }
  })
  store.on(LifecycleEvents.Delete, (entity: Entity<R>) => {
    try {
      invariant(
        incrementalStore.resource.id === entity.resource.id,
        "Subscriber Resource is for another Incremental Store",
      )
      incrementalStore.emit(
        LifecycleEvents.Delete,
        incrementalStore.resource,
        entity,
      )
      incrementalStore.emit(
        LifecycleEvents.Delta,
        incrementalStore.resource,
        entity,
      )
    } catch (error) {
      store.emit("error", error)
    }
  })
}
