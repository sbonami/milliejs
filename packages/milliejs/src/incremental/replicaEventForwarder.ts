import type { Entity, Resource } from "@milliejs/core"
import { LifecycleEvents } from "@milliejs/store-base"
import type { IncrementalStore } from "../incremental"
import type { PublisherActionEventInterface } from "../store/events"

export const replicaEventForwarder = <R extends Resource = Resource>(
  incrementalStore: IncrementalStore<R>,
  replicaStore: PublisherActionEventInterface<R>,
) => {
  Object.values(LifecycleEvents).forEach((event) => {
    replicaStore?.on(event, (entity: Entity<R>) => {
      incrementalStore.emit(event, incrementalStore.resource, entity)
    })
  })
}
