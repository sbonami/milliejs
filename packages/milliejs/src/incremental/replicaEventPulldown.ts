import invariant from "tiny-invariant"
import type { Entity, Query, Resource } from "@milliejs/core"
import { LifecycleEvents, SubscriberActionInterface } from "@milliejs/store-base"
import type { IncrementalStore } from "../incremental"

export const replicaEventPulldown = <R extends Resource = Resource>(
  incrementalStore: IncrementalStore<R>,
  store: SubscriberActionInterface,
) => {
  store.on(LifecycleEvents.Save, async (entity: Entity<R>) => {
    try {
      invariant(
        typeof incrementalStore.replicaStore !== "undefined",
        "Replica store cannot be undefined",
      )
      invariant(
        incrementalStore.resource.id === entity.resource.id,
        "Subscriber Resource is for another Incremental Store",
      )

      await incrementalStore.replicaStore.create(entity)
    } catch (error) {
      store.emit("error", error)
    }
  })

  store.on(LifecycleEvents.Delete, async (entityOrQuery: Entity<R> | Query) => {
    try {
      invariant(
        typeof incrementalStore.replicaStore !== "undefined",
        "Replica store cannot be undefined",
      )

      await incrementalStore.replicaStore.delete(entityOrQuery)
    } catch (error) {
      store.emit("error", error)
    }
  })
}
