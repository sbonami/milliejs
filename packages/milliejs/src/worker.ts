import type { StoreLifecycleInterface } from "@milliejs/store-base"

enum WorkerState {
  new = "new",
  starting = "starting",
  running = "running",
  exiting = "exiting",
  terminated = "terminated",
}

export abstract class Worker {
  private loop?: ReturnType<typeof setInterval>

  constructor(private _state: WorkerState = WorkerState.new) {
    process
      .on("SIGTERM", this.gracefulShutdown("SIGTERM"))
      .on("exit", this.gracefulShutdown("SIGTERM"))
  }

  protected abstract get connections(): Array<StoreLifecycleInterface>

  listen(callback?: () => void): Worker {
    this.loop = setInterval(async () => {
      try {
        if (this.state === WorkerState.new) {
          await this.init()
          if (callback) await callback()
        }
      } catch (error) {
        console.error({ error })
        await this.close()
        clearInterval(this.loop)
      }
    })

    return this
  }

  get state() {
    return this._state
  }

  private async init() {
    console.log("starting")
    this._state = WorkerState.starting

    await Promise.all(
      this.connections.map((connection) => connection.connect()),
    )

    console.log("running")
    this._state = WorkerState.running
  }

  async close(callback?: () => void) {
    console.log("exiting")
    this._state = WorkerState.exiting

    await Promise.all(
      this.connections.map((connection) => connection.disconnect()),
    )

    console.log("terminated")
    this._state = WorkerState.terminated

    if (callback) callback()
  }

  private gracefulShutdown(signal: string, code?: number) {
    return () => {
      console.debug(signal + " signal received: closing server")
      this.close(() => {
        console.debug("server closed")
        process.exit(code || 0)
      })
    }
  }
}

export default Worker
