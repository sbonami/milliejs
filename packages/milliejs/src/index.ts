import type { Entity, Query, Resource } from "@milliejs/core"
import type {
  PublisherActionInterface,
  SubscriberActionInterface,
} from "@milliejs/store-base"
import invariant from "tiny-invariant"
import { IncrementalStore, StoreConstructorSourceOptions } from "./incremental"

export type { Entity, Query, Resource }

class MillieJS {
  readonly stores: Record<Resource["id"], IncrementalStore> = {}

  registerResource<R extends Resource>(
    resource: R,
    replicaStore: PublisherActionInterface<R>,
    sourceInterfaces?: StoreConstructorSourceOptions<R>,
  ): IncrementalStore<R> {
    this.stores[resource.id] = new IncrementalStore<R>(resource, replicaStore)

    if (sourceInterfaces?.sourcePublisher)
      this.registerSourcePublisher(resource, sourceInterfaces.sourcePublisher)

    if (sourceInterfaces?.sourceSubscriber)
      this.registerSourceSubscriber(resource, sourceInterfaces.sourceSubscriber)

    return this.storeForResource<R>(resource)
  }

  registerSourcePublisher<R extends Resource>(
    resource: R,
    store: PublisherActionInterface<R>,
  ): IncrementalStore<R> {
    invariant(
      this.stores[resource.id],
      "Resource is not registered with this store",
    )

    this.stores[resource.id].sourcePublisher = store
    return this.storeForResource<R>(resource)
  }

  registerSourceSubscriber<R extends Resource>(
    resource: R,
    store: SubscriberActionInterface,
  ): IncrementalStore<R> {
    invariant(
      this.stores[resource.id],
      "Resource is not registered with this store",
    )

    this.stores[resource.id].sourceSubscriber = store
    return this.storeForResource<R>(resource)
  }

  private storeForResource<R extends Resource>(resource: R): IncrementalStore<R> {
    invariant(
      this.stores[resource.id],
      "Resource is not registered with this store",
    )

    return this.stores[resource.id] as IncrementalStore<R>
  }
}

export default MillieJS
