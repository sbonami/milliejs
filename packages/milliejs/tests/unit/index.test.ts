import type { Resource } from "@milliejs/core"
import { IncrementalStore } from "../../src/incremental"
import MillieJS from "../../src/index"
import { makeMockPublisherWithEvents } from "./mocks/publisher"
import { makeMockSubscriber } from "./mocks/subscriber"

jest.mock("../../src/incremental")

describe("MillieJS", () => {
  describe("registerResource", () => {
    it("generates a new IncrementalStore for the resource", () => {
      const mockResource: Resource = {
        id: "MOCK RESOURCE",
      }
      const mockReplicaStore = makeMockPublisherWithEvents()

      const millie = new MillieJS()
      millie.registerResource(mockResource, mockReplicaStore)

      // const spy = jest.spyOn(IncrementalStore, "prototype")

      expect(IncrementalStore).toHaveBeenCalledWith(mockResource, mockReplicaStore)
    })

    describe("when the resource is registered with a source publisher", () => {
      it("registers the source publisher for the resource", () => {
        const mockResource: Resource = {
          id: "MOCK RESOURCE",
        }
        const mockReplicaStore = makeMockPublisherWithEvents()
        const mockSourcePublisher = makeMockPublisherWithEvents()

        const millie = new MillieJS()
        const spy = jest.spyOn(millie, 'registerSourcePublisher')

        millie.registerResource(mockResource, mockReplicaStore, {
          sourcePublisher: mockSourcePublisher,
        })
        expect(spy).toHaveBeenCalledWith(mockResource, mockSourcePublisher)
      })
    })

    describe("when the resource is registered with a source subscriber", () => {
      it("registers the source subscriber for the resource", () => {
        const mockResource: Resource = {
          id: "MOCK RESOURCE",
        }
        const mockReplicaStore = makeMockPublisherWithEvents()
        const mockSourceSubscriber = makeMockSubscriber()

        const millie = new MillieJS()
        const spy = jest.spyOn(millie, 'registerSourceSubscriber')

        millie.registerResource(mockResource, mockReplicaStore, {
          sourceSubscriber: mockSourceSubscriber,
        })
        expect(spy).toHaveBeenCalledWith(mockResource, mockSourceSubscriber)
      })
    })

    it("returns the resource's IncrementalStore", () => {
      const MockedIncrementalStore = jest.mocked(IncrementalStore)
      const mockResource: Resource = {
        id: "MOCK RESOURCE",
      }
      const mockReplicaStore = makeMockPublisherWithEvents()
      const mockIncrementalStore = new IncrementalStore(mockResource, mockReplicaStore)
      MockedIncrementalStore.mockReturnValueOnce(mockIncrementalStore as any)

      const millie = new MillieJS()
      const response = millie.registerResource(mockResource, mockReplicaStore)
      expect(response).toBe(mockIncrementalStore)
    })
  })

  describe("registerSourcePublisher", () => {
    describe("when the resource has not yet been registered", () => {
      it("throws an error", () => {
        const mockResource: Resource = {
          id: "MOCK RESOURCE",
        }
        const mockSourcePublisher = makeMockPublisherWithEvents()

        const millie = new MillieJS()
        expect(() => {
          millie.registerSourcePublisher(mockResource, mockSourcePublisher)
        }).toThrowErrorMatchingSnapshot()
      })
    })

    it("sets the sourcePublisher on the registered resource's IncrementalStore", (done) => {
      expect.assertions(1)

      const MockedIncrementalStore = jest.mocked(IncrementalStore)
      const mockResource: Resource = {
        id: "MOCK RESOURCE",
      }
      const mockReplicaStore = makeMockPublisherWithEvents()
      const mockSourcePublisher = makeMockPublisherWithEvents()
      const mockIncrementalStore = new IncrementalStore(mockResource, mockReplicaStore)
      MockedIncrementalStore.mockReturnValueOnce(mockIncrementalStore as any)

      const millie = new MillieJS()
      millie.registerResource(mockResource, mockReplicaStore)

      Object.defineProperty(mockIncrementalStore, "sourcePublisher", {
        set: (newSourcePublisher: any) => {
          try {
            expect(newSourcePublisher).toBe(mockSourcePublisher)
            done()
          } catch (error) {
            done(error)
          }
        }
      })
      millie.registerSourcePublisher(mockResource, mockSourcePublisher)
    })

    it("returns the resource's IncrementalStore", () => {
      const MockedIncrementalStore = jest.mocked(IncrementalStore)
      const mockResource: Resource = {
        id: "MOCK RESOURCE",
      }
      const mockReplicaStore = makeMockPublisherWithEvents()
      const mockSourcePublisher = makeMockPublisherWithEvents()
      const mockIncrementalStore = new IncrementalStore(mockResource, mockReplicaStore)
      MockedIncrementalStore.mockReturnValueOnce(mockIncrementalStore as any)

      const millie = new MillieJS()
      millie.registerResource(mockResource, mockReplicaStore)

      const response = millie.registerSourcePublisher(mockResource, mockSourcePublisher)
      expect(response).toBe(mockIncrementalStore)
    })
  })

  describe("registerSourceSubscriber", () => {
    describe("when the resource has not yet been registered", () => {
      it("throws an error", () => {
        const mockResource: Resource = {
          id: "MOCK RESOURCE",
        }
        const mockSourceSubscriber = makeMockSubscriber()

        const millie = new MillieJS()
        expect(() => {
          millie.registerSourceSubscriber(mockResource, mockSourceSubscriber)
        }).toThrowErrorMatchingSnapshot()
      })
    })

    it("sets the sourceSubscriber on the registered resource's IncrementalStore", (done) => {
      expect.assertions(1)

      const MockedIncrementalStore = jest.mocked(IncrementalStore)
      const mockResource: Resource = {
        id: "MOCK RESOURCE",
      }
      const mockReplicaStore = makeMockPublisherWithEvents()
      const mockSourceSubscriber = makeMockSubscriber()
      const mockIncrementalStore = new IncrementalStore(mockResource, mockReplicaStore)
      MockedIncrementalStore.mockReturnValueOnce(mockIncrementalStore as any)

      const millie = new MillieJS()
      millie.registerResource(mockResource, mockReplicaStore)

      Object.defineProperty(mockIncrementalStore, "sourceSubscriber", {
        set: (newSourceSubscriber: any) => {
          try {
            expect(newSourceSubscriber).toBe(mockSourceSubscriber)
            done()
          } catch (error) {
            done(error)
          }
        }
      })
      millie.registerSourceSubscriber(mockResource, mockSourceSubscriber)
    })

    it("returns the resource's IncrementalStore", () => {
      const MockedIncrementalStore = jest.mocked(IncrementalStore)
      const mockResource: Resource = {
        id: "MOCK RESOURCE",
      }
      const mockReplicaStore = makeMockPublisherWithEvents()
      const mockSourceSubscriber = makeMockSubscriber()
      const mockIncrementalStore = new IncrementalStore(mockResource, mockReplicaStore)
      MockedIncrementalStore.mockReturnValueOnce(mockIncrementalStore as any)

      const millie = new MillieJS()
      millie.registerResource(mockResource, mockReplicaStore)

      const response = millie.registerSourceSubscriber(mockResource, mockSourceSubscriber)
      expect(response).toBe(mockIncrementalStore)
    })
  })
})
