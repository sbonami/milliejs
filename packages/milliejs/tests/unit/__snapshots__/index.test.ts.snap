// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MillieJS registerSourcePublisher when the resource has not yet been registered throws an error 1`] = `"Invariant failed: Resource is not registered with this store"`;

exports[`MillieJS registerSourceSubscriber when the resource has not yet been registered throws an error 1`] = `"Invariant failed: Resource is not registered with this store"`;
